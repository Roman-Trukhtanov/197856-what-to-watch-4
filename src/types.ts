export type Time = number | string;

export type VideoTypes = {
  [key: string]: string;
}

export type Action = {
  [key: string]: string;
}

export type ActionObject<T> = {
  type: string;
  payload: T;
}

export type Noop = () => void;

export type FavoriteStatus = 0 | 1;

export interface Review {
  rating: number;
  comment: string;
}

export interface PreviewVideo {
  className?: string;
  src: string;
  type: string;
  isAutoPlay?: boolean;
  isLoop: boolean;
  isMute: boolean;
}
export interface FullVideo {
  className: string;
  src: string;
  type: string;
  isAutoPlay?: boolean;
  isLoop: boolean;
  isMute: boolean;
}

export interface Movie {
  id: number;
  title: string;
  genre: string;
  coverSrc: string;
  bgColor: string;
  bigPosterSrc: string;
  previewImgSrc: string;
  isFavorite: boolean;
  fullVideo: FullVideo;
  previewVideo: PreviewVideo;
  details: {
    rate: number;
    releaseYear: number;
    ratingCount: number;
    level: string;
    director: string;
    runTime: number;
    description: string;
    starring: string[];
  };
}

export interface Comment {
  id: number;
  user: {
    id: number;
    name: string;
  };
  rating: number;
  comment: string;
  date: string;
  dateText: string;
}

export interface User {
  id: number;
  name: string;
  email: string;
  avatarSrc: string;
}

export interface UserData {
  email: string;
  password: string;
}

export interface ImgBgStyle {
  display: string;
  backgroundColor: string;
}
