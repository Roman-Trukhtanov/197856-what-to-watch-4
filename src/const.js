export const MAX_GENRES_AMOUNT = 10;

export const MAX_VISIBLE_MOVIES_COUNT = 8;

export const OVERVIEW_STARRING_COUNT = 4;

export const MAX_SLICED_MOVIES = 4;

export const DEFAULT_SELECTED_MOVIE_ID = 1;

export const DEFAULT_COLLECTION_NUMBER = 1;

export const MINUTES_IN_HOUR = 60;
export const SECONDS_IN_MINUTE = 60;

export const REVIEW_STARS_AMOUNT = 5;

export const DEFAULT_RATING = 3;

export const RATING_MULTIPLIER = 2;

export const DEFAULT_BG_COLOR = `#250505`;

export const DEFAULT_VIDEO_TYPE = `video/mp4`;

export const ScreenType = {
  MAIN: `main`,
  MOVIE: `movie`,
  PLAYER: `player`,
  SIGN_IN: `sign-in`,
  ADD_REVIEW: `add-review`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const DEFAULT_GENRE = `All Genres`;

export const TabType = {
  OVERVIEW: `overview`,
  DETAILS: `details`,
  REVIEWS: `reviews`,
};

export const monthsNames = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`,
];

export const VideoType = {
  MP4: `video/mp4`,
  WEBM: `video/webm`,
};


export const ReviewTextLimit = {
  MIN: 50,
  MAX: 400,
};

export const ReviewMessage = {
  SUCCESS: `Your comment has been posted successfully`,
  ERROR: `An error occurred while posting a comment, please try again ...`,
};
