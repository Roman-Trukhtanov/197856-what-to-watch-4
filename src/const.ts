import {VideoTypes} from "./types";

export const MAX_GENRES_AMOUNT = 10;

export const MAX_VISIBLE_MOVIES_COUNT = 8;

export const OVERVIEW_STARRING_COUNT = 4;

export const MAX_SLICED_MOVIES = 4;

export const DEFAULT_SELECTED_MOVIE_ID = 1;

export const DEFAULT_COLLECTION_NUMBER = 1;

export const MINUTES_IN_HOUR = 60;
export const SECONDS_IN_MINUTE = 60;

export const REVIEW_STARS_AMOUNT = 5;

export const DEFAULT_RATING = 3;

export const RATING_MULTIPLIER = 2;

export const DEFAULT_BG_COLOR = `#250505`;

export enum ScreenType {
  FILMS = `films`,
  PLAYER =`player`,
  REVIEW = `review`,
}

export enum AppRoute {
  ROOT = `/`,
  SIGN_IN = `/login`,
  PLAYER = `/player/:id`,
  MOVIE = `/films/:id`,
  REVIEW = `/films/:id/review`,
  MY_LIST = `/mylist`,
}

export enum HistoryAction {
  PUSH = `PUSH`,
}

export enum AuthorizationStatus {
  AUTH = `AUTH`,
  NO_AUTH = `NO_AUTH`,
}

export const DEFAULT_GENRE = `All Genres`;

export enum TabType {
  OVERVIEW = `overview`,
  DETAILS = `details`,
  REVIEWS = `reviews`,
}

export const monthsNames: Array<string> = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`,
];

export const DEFAULT_VIDEO_TYPE = `video/mp4`;

export const VideoType: VideoTypes = {
  MP4: `video/mp4`,
  WEBM: `video/webm`,
};

export enum ReviewTextLimit {
  MIN = 50,
  MAX = 400,
}

export enum ReviewMessage {
  SUCCESS = `Your comment has been posted successfully`,
  ERROR = `An error occurred while posting a comment, please try again ...`,
}

export const PAGE_NOT_FOUND_TEXT = `404: Page not found...`;
