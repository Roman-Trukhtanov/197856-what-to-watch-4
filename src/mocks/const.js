export const MAX_GENRES_AMOUNT = 10;

export const MAX_VISIBLE_MOVIES_COUNT = 8;

export const OVERVIEW_STARRING_COUNT = 4;

export const ScreenType = {
  MAIN: `main`,
  MOVIE: `movie`,
};

export const GenreType = {
  ALL_GENRES: `All Genres`,
  COMEDIES: `Comedies`,
  CRIME: `Crime`,
  DOCUMENTARY: `Documentary`,
  DRAMAS: `Dramas`,
  THRILLERS: `Thrillers`,
  HORROR: `Horror`,
  ROMANCE: `Romance`,
  FAMILY: `Family`,
  FANTASY: `Fantasy`,
};

export const TabType = {
  OVERVIEW: `overview`,
  DETAILS: `details`,
  REVIEWS: `reviews`,
};
