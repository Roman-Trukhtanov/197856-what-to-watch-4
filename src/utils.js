import {
  DEFAULT_BG_COLOR,
  MAX_VISIBLE_MOVIES_COUNT,
  MINUTES_IN_HOUR,
  SECONDS_IN_MINUTE
} from "./const";

export const extend = (currentObject, newObject) => {
  return Object.assign({}, currentObject, newObject);
};

export const spliceMovies = (
    collectionNumber = 1,
    movies,
    visibleMoviesCount = MAX_VISIBLE_MOVIES_COUNT
) => {
  return [...movies].splice(0, visibleMoviesCount * collectionNumber);
};

export const checkVisibleMovies = (movies, movieCollectionNumber) => {
  return MAX_VISIBLE_MOVIES_COUNT * movieCollectionNumber < movies.length;
};

export const getStringTime = (secValue = 0) => {
  let hours = Math.floor(secValue / SECONDS_IN_MINUTE / MINUTES_IN_HOUR);
  let minutes = Math.floor(secValue / SECONDS_IN_MINUTE % MINUTES_IN_HOUR);
  let seconds = Math.floor(secValue % SECONDS_IN_MINUTE);

  if (minutes < 10) {
    minutes = `0${minutes}`;
  }
  if (hours < 10) {
    hours = `0${hours}`;
  }
  if (seconds < 10) {
    seconds = `0${seconds}`;
  }

  return `${hours}:${minutes}:${seconds}`;
};

export const scrollToTop = () => {
  window.scrollTo({
    top: 0,
    behavior: `smooth`
  });
};

export const getImgBgStyle = (bgColor) => ({
  display: `block`,
  backgroundColor: bgColor || DEFAULT_BG_COLOR,
});

export const shuffleArray = (array) => {
  const copiedArray = [...array];

  for (let i = copiedArray.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    const temp = copiedArray[j];
    copiedArray[j] = copiedArray[i];
    copiedArray[i] = temp;
  }
  return copiedArray;
};
