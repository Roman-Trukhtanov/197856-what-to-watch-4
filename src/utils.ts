import {
  DEFAULT_BG_COLOR,
  MAX_VISIBLE_MOVIES_COUNT,
  MINUTES_IN_HOUR,
  SECONDS_IN_MINUTE
} from "./const";
import {ImgBgStyle, Movie, Time} from "./types";

export const extend = (currentObject: object, newObject: object): object => {
  return Object.assign({}, currentObject, newObject);
};

export const noop = () => {
  return;
};

export const spliceMovies = (
    collectionNumber = 1,
    movies: Movie[],
    visibleMoviesCount: number = MAX_VISIBLE_MOVIES_COUNT
): Movie[] => {
  return [...movies].splice(0, visibleMoviesCount * collectionNumber);
};

export const checkVisibleMovies = (
    movies: Movie[],
    movieCollectionNumber: number
): boolean => {
  return MAX_VISIBLE_MOVIES_COUNT * movieCollectionNumber < movies.length;
};

export const getStringTime = (secValue = 0): string => {
  let hours: Time = Math.floor(secValue / SECONDS_IN_MINUTE / MINUTES_IN_HOUR);
  let minutes: Time = Math.floor(secValue / SECONDS_IN_MINUTE % MINUTES_IN_HOUR);
  let seconds: Time = Math.floor(secValue % SECONDS_IN_MINUTE);

  if (minutes < 10) {
    minutes = `0${minutes}`;
  }
  if (hours < 10) {
    hours = `0${hours}`;
  }
  if (seconds < 10) {
    seconds = `0${seconds}`;
  }

  return `${hours}:${minutes}:${seconds}`;
};

export const scrollToTop = (): void => {
  window.scrollTo({
    top: 0,
    behavior: `smooth`
  });
};

export const getImgBgStyle = (bgColor: string): ImgBgStyle => ({
  display: `block`,
  backgroundColor: bgColor || DEFAULT_BG_COLOR,
});

export const shuffleMovies = (Movies: Movie[]): Movie[] => {
  const copiedArray: Movie[] = [...Movies];

  for (let i = copiedArray.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    const temp = copiedArray[j];
    copiedArray[j] = copiedArray[i];
    copiedArray[i] = temp;
  }

  return copiedArray;
};

export const reloadPage = (): void => {
  window.location.reload();
};
