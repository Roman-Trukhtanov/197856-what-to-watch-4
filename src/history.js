import {createBrowserHistory} from "history";
import {AppRoute, HistoryAction} from "./const";

const history = createBrowserHistory();

export const historyGoBack = () => {
  const historyAction = history.action;

  if (historyAction === HistoryAction.PUSH) {
    history.goBack();
  } else {
    history.push(AppRoute.ROOT);
  }
};

export default history;
