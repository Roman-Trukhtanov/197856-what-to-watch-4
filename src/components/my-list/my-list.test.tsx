import React from "react";
import {Router} from "react-router-dom";
import renderer from "react-test-renderer";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import MyList from "./my-list";
import history from "../../history";
import NameSpace from "../../reducer/name-space";
import {AuthorizationStatus} from "../../const";
import {Movie} from "../../types";
import {noop} from "../../utils";

const mockStore = configureStore([]);

const mockMovieData: Movie[] = [{
  id: 1,
  title: `Snatch`,
  genre: `Comedy`,
  bgColor: `#FDFDFC`,
  coverSrc: `https://some-link`,
  bigPosterSrc: `https://some-link`,
  previewImgSrc: `https://some-link`,
  isFavorite: false,
  fullVideo: {
    className: `player__video`,
    src: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    type: `video/mp4`,
    isAutoPlay: true,
    isLoop: false,
    isMute: false,
  },
  previewVideo: {
    src: `https://some-link`,
    type: `video/mp4`,
    isAutoPlay: true,
    isLoop: true,
    isMute: true,
  },
  details: {
    rate: 3,
    releaseYear: 2000,
    ratingCount: 100,
    level: `Bad`,
    description: `Description`,
    director: `Guy Ritchie`,
    runTime: 104,
    starring: [`Jason Statham`, `Brad Pitt`, `Benicio Del Toro`],
  },
}];

describe(`MyList component`, () => {
  it(`Render MyList`, () => {
    const store = mockStore({
      [NameSpace.DATA]: {
        movies: mockMovieData,
        promoMovie: mockMovieData[0],
        favoriteMovies: mockMovieData,
      },
      [NameSpace.USER]: {
        authorizationStatus: AuthorizationStatus.NO_AUTH,
        user: {
          id: 0,
          name: ``,
          email: ``,
          avatarSrc: ``,
        },
      }
    });

    const tree = renderer
      .create(
          <Provider store={store}>
            <Router
              history={history}
            >
              <MyList
                favoriteMovies={mockMovieData}
                onMovieCardTitleClick={noop}
              />
            </Router>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
