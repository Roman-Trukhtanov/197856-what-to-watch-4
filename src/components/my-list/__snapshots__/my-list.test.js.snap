// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyList component Render MyList 1`] = `
<div
  className="user-page"
>
  <header
    className="page-header movie-card__head"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link"
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <h1
      className="page-title user-page__title"
    >
      My list
    </h1>
    <div
      className="user-block"
    >
      <a
        className="user-block__link"
        href="/login"
        onClick={[Function]}
      >
        Sign in
      </a>
    </div>
  </header>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <div
      className="catalog__movies-list"
    >
      <div
        className="catalog__movies-list"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
            onClick={[Function]}
          >
            <img
              alt="Snatch"
              height="175"
              src="https://some-link"
              style={
                Object {
                  "backgroundColor": "#FDFDFC",
                  "display": "block",
                }
              }
              width="280"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="movie-page.html"
              onClick={[Function]}
            >
              Snatch
            </a>
          </h3>
        </article>
      </div>
    </div>
  </section>
  <footer
    className="page-footer"
  >
    <div
      className="logo"
    >
      <a
        className="logo__link logo__link--light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo__letter logo__letter--1"
        >
          W
        </span>
        <span
          className="logo__letter logo__letter--2"
        >
          T
        </span>
        <span
          className="logo__letter logo__letter--3"
        >
          W
        </span>
      </a>
    </div>
    <div
      className="copyright"
    >
      <p>
        Â© 2020 What to watch Ltd.
      </p>
    </div>
  </footer>
</div>
`;
